Creates a way to test private APIs in the Flutter package.

gspencergoog opened the pull request:
## Description

This adds a way to test private APIs in the flutter package by copying the appropriate private interfaces into a temporary directory, and running them there by creating a new library using `part of`.

## Related Issues

 - https://github.com/dart-lang/sdk/issues/40146 because this work needed to happen in order to eliminate the `part of` usage in flutter's tests.

## Tests

 - This removes a test that uses part to bring in the actual APIs from the framework, and replaces it with one that copies those private portions into a temporary directory.

## Breaking Change

- [X] No, no existing tests failed, so this is *not* a breaking change.

gspencergoog committed:
Add a new way to test private interfaces

gspencergoog committed:
Add to test.dart

gspencergoog committed:
Update pubspecs

Hixie commented on dev/manual_tests/pubspec.yaml:
```diff
@@ -18,23 +18,49 @@ dev_dependencies:
   test_api: 0.2.19-nullsafety # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
+  watcher: 0.9.7+15 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
+  yaml: 2.2.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
 
```
let's pin mockito to 4.1.1 since we're planning on removing it entirely anyway (right? cc @jonahwilliams)
(I'm assuming it's the cause of all these new dependencies.)

jonahwilliams replied:
mockito still exists in one test, the one that uses private APIs. Once it is removed from there it will be gone from `package:flutter`, but still used elsewhere in the SDK

Hixie replied:
Are we not planning on removing it from everywhere?

gspencergoog replied:
But is it OK to pin it to 4.1.1 for `flutter_tool`?

jonahwilliams replied:
We're not going to migrate to NNBD in the tool for quite a while, so it should be safe for now

Hixie commented on packages/flutter/test_private/README.md:
```diff
@@ -0,0 +1,51 @@
+
+One limitation is that the copied code needs to be separable enough to be
+copied, so it needs to be in its own separate files, or extracted out of a
+larger file by the script (not a method currently employed).
```
let's not encourage that second path if we don't need it... requiring that this approach only be used by entirely separable code seems like a good place to start.

gspencergoog replied:
Good point.  Removed.

Hixie commented on packages/flutter/test_private/pubspec.yaml:
```diff
@@ -0,0 +1,24 @@
+
+dependencies:
+  # To update these, use "flutter update-packages --force-upgrade".
+  args: 1.6.0
```
you don't actually use args, it seems

gspencergoog replied:
Right. Removed. (I would if there was a nullsafety version.)

Hixie commented on packages/flutter/test_private/bin/test_private.dart:
```diff
@@ -0,0 +1,208 @@
+///
+/// The pubspec file is copied to the root of the test area too, but renamed to
+/// "pubspec.yaml".
+class TestCase extends Object {
```
"extends Object" is redundant

gspencergoog replied:
Yeah, I know, but there were some weird analysis problems with non-nullable code that it was fixing.  It appears to not be needed now though, so maybe it was a bug in the analyzer that got fixed.

Hixie commented on packages/flutter/test_private/bin/test_private.dart:
```diff
@@ -0,0 +1,208 @@
+      // .dart_tool, which contains a (non-hidden) .json file.
+      continue;
+    }
+    if (entity is File && path.basename(entity.path).endsWith('.json')) {
```
should it end with `_test.json`?

gspencergoog replied:
Good idea.  Done.

Hixie commented:
LGTM, but you'll have to find a way to make the analyzer work in this regime and it looks like the script didn't quite work in CI.

gspencergoog committed:
Review Changes, pin Mockito to 4.1.1

gspencergoog committed:
Add .dartignore to test directory, because imports will always be missing.

gspencergoog reviewed (commented):
I also added a `.dartignore` to the test directory, since the `part` directives in the test will always be invalid paths, and so won't pass analysis.

I'll probably need to add an analysis pass to the private test script to make sure there aren't analysis issues in the test(s).

goderbauer commented on packages/flutter/test_private/README.md:
```diff
@@ -0,0 +1,50 @@
+making a copy of the code under test.
+
+The test script `bin/test_private.dart` tests private interfaces by copying the
+code under test into a temporary workspace.  The test is then free to make the
```
nit: double space

gspencergoog replied:
Fixed.

goderbauer reviewed (approved):
LGTM

goderbauer commented:
Looks like various checks are failing though...

gspencergoog committed:
Skip test code in regular analysis, add analysis step to private tests

gspencergoog commented:
OK, I added an analysis step to the running of the private tests in the tmpdir, and changed the name of the test file in the tree so that it wouldn't get analyzed, since analyzing it out of context will always fail.
